sw/device/lib/base/mmio.h

```diff
@@ -234,6 +234,62 @@ inline void mmio_region_nonatomic_set_mask32(mmio_region_t base,
   mmio_region_write32(base, offset, value);
 }
 
+/**
+ * Masked Register field.
+ */
+typedef struct mmio_region_field32 {
+  uint32_t mask;  /**< The field mask. */
+  uint32_t index; /**< The field position within the a register. */
+  uint32_t value; /**< The field value to be written. */
+} mmio_region_field32_t;
+
+/**
+ * Sets the |field| from the MMIO region |base| at the given offset.
+ *
+ * This function performs a non-atomic read-write-modify operation on a
+ * MMIO region.
+ *
+ * @param base the region to set the field in.
+ * @param offset the offset to set the field at, in bytes.
+ * @param field field within selected register field to be set.
+ */
+inline void mmio_region_nonatomic_set_field32(mmio_region_t base,
+                                              ptrdiff_t offset,
+                                              mmio_region_field32_t field) {
+  uint32_t register_value = mmio_region_read32(base, offset);
+  register_value &= ~(field.mask << field.index);
+  register_value |= (field.value & field.mask) << field.index;
+  mmio_region_write32(base, offset, register_value);
+}
+
+/**
+ * Convienience macro for mmio_region_nonatomic_set_fields32.
+ */
+#define MMIO_REGION_NONATOMIC_SET_FIELDS32(base, offset, fields) \
+  mmio_region_nonatomic_set_fields32(base, offset, &fields,      \
+                                     sizeof(fields) / sizeof(fields[0]))
+
+/**
+ * Sets the |fields| from the MMIO region |base| at the given offset.
+ *
+ * This function performs a non-atomic read-write-modify operation on a
+ * MMIO region.
+ *
+ * @param base the region to set the field in.
+ * @param offset the offset to set the field at, in bytes.
+ * @param fields fields within selected register to be set.
+ */
+inline void mmio_region_nonatomic_set_fields32(
+    mmio_region_t base, ptrdiff_t offset, const mmio_region_field32_t *fields,
+    int num_fields) {
+  uint32_t register_value = mmio_region_read32(base, offset);
+  for (int i = 0; i < num_fields; ++i) {
+    register_value &= ~(fields[i].mask << fields[i].index);
+    register_value |= (fields[i].value & fields[i].mask) << fields[i].index;
+  }
+  mmio_region_write32(base, offset, register_value);
+}
```

Comment:
Does this compile to anything at all reasonable? I'm not convinced this code is the best idea.

Reply:
The honest answer is, I haven't thoroughly examined it yet. The idea was to produce something that people can have a look at and decide whether it would be useful or not so much.

Reply:
This code no longer exists.